### PR TITLE
New relocation-generating program

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

Added a new relocation-generating program `Fado`, written by me, and changed the Makefile to generate all the reloc file dependencies and build them automatically.

Useful information about Fado:

- Automatically compatible with both IDO and GCC.
- Handles both single and multi-file overlays correctly.
- Prints the standard .ovl section format, with the relocs divided by section. It will also print the name the associated variable if it can find it in the ELF (for IDO, this is only if it is not static, whereas GCC sometimes seems to retain all of the variable names).

I used valgrind to test making the reloc file for ovl_kaleido_scope, one of the larger and more complicated overlays, with both fado and ZAPD:

Fado:
```
==16833== HEAP SUMMARY:
==16833==     in use at exit: 0 bytes in 0 blocks
==16833==   total heap usage: 138 allocs, 138 frees, 205,195 bytes allocated
```

ZAPD:
```
==17052== HEAP SUMMARY:
==17052==     in use at exit: 35,066 bytes in 1,736 blocks
==17052==   total heap usage: 316,758 allocs, 315,022 frees, 10,588,407 bytes allocated
```

So ZAPD uses 50 times as much memory and almost 2300 times as many allocations. Speedwise, processing every reloc in OoT,

Fado:
```
real    0m0.446s
user    0m0.332s
sys     0m0.149s
```

ZAPD:
```
real    0m4.122s
user    0m3.685s
sys     0m0.482s
```

i.e. approximately 10 times faster. I have to admit I was pretty staggered by these figures.

This also fixes the annoying `Error 1 (ignored)`s (suppressed in MM, still shown in OoT) with proper dependencies for the relocs.

This PR is not yet complete, in that when we move to Fado, we will no longer need all the .cfg files. I elected not to delete them yet so as not to obscure what this PR actually does behind a wall of less significant changed files: once everyone's satisfied, I can remove them.

### CODE REVIEW

**Review by jpburnett — APPROVED** [review]
Gotta go FAST 🟦 (there was no blue circle...)

**Comment by EllipticEllipsis** [comment]
I'm going to mark this as draft until OoT works out what to do with this, since it might benefit MM too.

File: Makefile [hunk 1]
```diff
@@ -334,13 +350,13 @@ build/data/%.o: data/%.s
 build/src/overlays/%.o: src/overlays/%.c
 	$(CC_CHECK) $<
 	$(CC) -c $(CFLAGS) $(MIPS_VERSION) $(OPTFLAGS) -o $@ $<
-	$(OBJDUMP_CMD)
-# TODO: `() || true` is currently necessary to suppress `Error 1 (ignored)` make warnings caused by `test`, but this will go away if 
-# 	the following is moved to a separate rule that is only run once when all the required objects have been compiled. 
-	$(ZAPD) bovl -eh -i $@ -cfg $< --outputpath $(@D)/$(notdir $(@D))_reloc.s
-	(test -f $(@D)/$(notdir $(@D))_reloc.s && $(AS) $(ASFLAGS) $(@D)/$(notdir $(@D))_reloc.s -o $(@D)/$(notdir $(@D))_reloc.o) || true
+	@$(OBJDUMP) -d $@ > $(@:.o=.s)
```
Comment (hensldm) [inline Makefile:353]: Yay, no more `|| true`

**Review by hensldm — APPROVED** [review]
I can't speak to the tool refactors, but MM stuff looks good. Test locally and it seems fine and snappy.

**Review by AngheloAlf — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**